# Galois Field Demonstrations

This notebook demonstrates arithmetic in $GF(2^m)$ for various $m$, including addition, multiplication, and polynomial representations. We construct and display elements of finite fields, showing their integer, binary, primitive element, and polynomial forms.

## Setup and Imports

We import the `GF` class, which implements arithmetic in $GF(p^m)$, and set up the Python environment for our coding theory demonstrations.

In [1]:
import numpy as np
import sys
import os
# Ensure src is in the Python path for module imports
SRC_PATH = os.path.abspath(os.path.join(os.getcwd(), '../src'))
if SRC_PATH not in sys.path:
    sys.path.insert(0, SRC_PATH)
from codes.gf import GF
from codes.bch import BCH
from utils import encode, poly_to_string

## Constructing $GF(8)$

We construct the field $GF(8)$ using the primitive polynomial $x^3 + x + 1$. In binary, this polynomial is represented as `[1, 1, 0, 1]` (corresponding to $1 + x + x^3$).

The field has $2^3 = 8$ elements.

In [2]:
# Example: GF(8) with primitive polynomial x^3 + x + 1
gf8 = GF(2, 3, [1, 1, 0, 1])  # binary polynomial: 1 + x + x^3

## Table of All Elements in $GF(8)$

Below, each element of $GF(8)$ is shown in four forms:
- **Integer representation** (0 to 7)
- **Binary representation** (as a 3-bit string)
- **Power of the primitive element $\alpha$** (with $\alpha = x$), or `0` for the zero element
- **Polynomial representation** (in terms of $\alpha$)

In [3]:
gf8.print_table()

Int |   Binary |    α^k |           Polynomial
---------------------------------------------
  0 |      000 |      0 |                    0
  1 |      001 |      1 |                    1
  2 |      010 |     α¹ |                    α
  4 |      100 |     α² |                   α²
  3 |      011 |     α³ |                α + 1
  6 |      110 |     α⁴ |               α² + α
  7 |      111 |     α⁵ |           α² + α + 1
  5 |      101 |     α⁶ |               α² + 1


## Addition Example

We demonstrate addition in $GF(8)$ by adding $\alpha$ and $\alpha^2$. The result is $\alpha + \alpha^2$.

In [4]:
a = [0,1,0]
b = [0,0,1]

result = gf8.add(a, b)  # Should return [0, 1, 1] (binary addition)
print(f"Addition result: {result}")


Addition result: 6


Note that the result from the addition is the integer 6. In the table above, $\alpha$ is $010$ in binary representation and $\alpha^2$ is $100$. Their sum in binary is $110$, which is the integer 6. We can print this result in polynomial format using the `poly_str` method.

In [5]:
print(gf8.poly_str(result))

α² + α


## Multiplication Example

We demonstrate multiplication in $GF(8)$ by multiplying $\alpha$ and $\alpha^2$. The result is $\alpha^3$, which is reduced modulo the primitive polynomial $x^3 + x + 1$.

In [6]:
prod_result = gf8.mul(a, b)
print("α * α^2 =", gf8.poly_str(prod_result))

α * α^2 = α + 1


## Tables for Field Extensions of $GF(2)$

We construct and display tables for $GF(4)$, $GF(8)$, $GF(16)$, $GF(32)$, and $GF(64)$ using different primitive polynomials.

Table for $GF(4)= GF(2)(\alpha)$ where $\alpha^2 + \alpha + 1 = 0$

In [7]:
gf4 = GF(2,2,[1,1,1])  # α^2 + α + 1
gf4.print_table()

Int |   Binary |    α^k |           Polynomial
---------------------------------------------
  0 |       00 |      0 |                    0
  1 |       01 |      1 |                    1
  2 |       10 |     α¹ |                    α
  3 |       11 |     α² |                α + 1


Table for $GF(8)= GF(2)(\alpha)$ where $\alpha^3 + \alpha + 1 = 0$

In [8]:
gf8.print_table()

Int |   Binary |    α^k |           Polynomial
---------------------------------------------
  0 |      000 |      0 |                    0
  1 |      001 |      1 |                    1
  2 |      010 |     α¹ |                    α
  4 |      100 |     α² |                   α²
  3 |      011 |     α³ |                α + 1
  6 |      110 |     α⁴ |               α² + α
  7 |      111 |     α⁵ |           α² + α + 1
  5 |      101 |     α⁶ |               α² + 1


Table for $GF(16)= GF(2)(\alpha)$ where $\alpha^4 + \alpha^3 + 1 = 0$

In [12]:
gf16 = GF(2,4,[1,0,0,1,1])  # α^4 + α^3 + 1
gf16.print_table()

Int |   Binary |    α^k |           Polynomial
---------------------------------------------
  0 |     0000 |      0 |                    0
  1 |     0001 |      1 |                    1
  2 |     0010 |     α¹ |                    α
  4 |     0100 |     α² |                   α²
  8 |     1000 |     α³ |                   α³
  9 |     1001 |     α⁴ |               α³ + 1
 11 |     1011 |     α⁵ |           α³ + α + 1
 15 |     1111 |     α⁶ |      α³ + α² + α + 1
  7 |     0111 |     α⁷ |           α² + α + 1
 14 |     1110 |     α⁸ |          α³ + α² + α
  5 |     0101 |     α⁹ |               α² + 1
 10 |     1010 |    α¹⁰ |               α³ + α
 13 |     1101 |    α¹¹ |          α³ + α² + 1
  3 |     0011 |    α¹² |                α + 1
  6 |     0110 |    α¹³ |               α² + α
 12 |     1100 |    α¹⁴ |              α³ + α²


Table for $GF(32)= GF(2)(\alpha)$ where $\alpha^5 + \alpha^2 + 1 = 0$

In [13]:
gf32 = GF(2,5,[1,0,1,0,0,1])  # α^5 + α^2 + 1
gf32.print_table()

Int |   Binary |    α^k |           Polynomial
---------------------------------------------
  0 |    00000 |      0 |                    0
  1 |    00001 |      1 |                    1
  2 |    00010 |     α¹ |                    α
  4 |    00100 |     α² |                   α²
  8 |    01000 |     α³ |                   α³
 16 |    10000 |     α⁴ |                   α⁴
  5 |    00101 |     α⁵ |               α² + 1
 10 |    01010 |     α⁶ |               α³ + α
 20 |    10100 |     α⁷ |              α⁴ + α²
 13 |    01101 |     α⁸ |          α³ + α² + 1
 26 |    11010 |     α⁹ |          α⁴ + α³ + α
 17 |    10001 |    α¹⁰ |               α⁴ + 1
  7 |    00111 |    α¹¹ |           α² + α + 1
 14 |    01110 |    α¹² |          α³ + α² + α
 28 |    11100 |    α¹³ |         α⁴ + α³ + α²
 29 |    11101 |    α¹⁴ |     α⁴ + α³ + α² + 1
 31 |    11111 |    α¹⁵ | α⁴ + α³ + α² + α + 1
 27 |    11011 |    α¹⁶ |      α⁴ + α³ + α + 1
 19 |    10011 |    α¹⁷ |           α⁴ + α + 1
  3 |    00011

Table for $GF(64)= GF(2)(\alpha)$ where $\alpha^6 + \alpha + 1 = 0$

In [14]:
gf64 = GF(2,6,[1,1,0,0,0,0,1])  # α^6 + α + 1
gf64.print_table()

Int |   Binary |    α^k |           Polynomial
---------------------------------------------
  0 |   000000 |      0 |                    0
  1 |   000001 |      1 |                    1
  2 |   000010 |     α¹ |                    α
  4 |   000100 |     α² |                   α²
  8 |   001000 |     α³ |                   α³
 16 |   010000 |     α⁴ |                   α⁴
 32 |   100000 |     α⁵ |                   α⁵
  3 |   000011 |     α⁶ |                α + 1
  6 |   000110 |     α⁷ |               α² + α
 12 |   001100 |     α⁸ |              α³ + α²
 24 |   011000 |     α⁹ |              α⁴ + α³
 48 |   110000 |    α¹⁰ |              α⁵ + α⁴
 35 |   100011 |    α¹¹ |           α⁵ + α + 1
  5 |   000101 |    α¹² |               α² + 1
 10 |   001010 |    α¹³ |               α³ + α
 20 |   010100 |    α¹⁴ |              α⁴ + α²
 40 |   101000 |    α¹⁵ |              α⁵ + α³
 19 |   010011 |    α¹⁶ |           α⁴ + α + 1
 38 |   100110 |    α¹⁷ |          α⁵ + α² + α
 15 |   001111